File: ruby/2015/day03b/day03b.rb
#!/usr/bin/env ruby

require 'set'

def usage
    progname = File.basename($0)
    puts "usage: #{progname} <input file>"
    exit 1
end

def process(filename)
    santa = [0, 0]
    robo_santa = [0, 0]
    positions = Set.new()
    positions.add(santa)
    i = 0
    File.foreach(filename) do |line|
        line.each_char do |ch|
            santa_move = (i % 2) == 0
            case ch
            when '^'
                santa_move ? santa = [santa[0], santa[1] + 1] : robo_santa = [robo_santa[0], robo_santa[1] + 1]
            when 'v'
                santa_move ? santa = [santa[0], santa[1] - 1] : robo_santa = [robo_santa[0], robo_santa[1] - 1]
            when '<'
                santa_move ? santa = [santa[0] - 1, santa[1]] : robo_santa = [robo_santa[0] - 1, robo_santa[1]]
            when '>'
                santa_move ? santa = [santa[0] + 1, santa[1]] : robo_santa = [robo_santa[0] + 1, robo_santa[1]]
            end
            santa_move ? positions.add(santa) : positions.add(robo_santa)
            i += 1
        end
    end
    positions.size
end

def main
    usage if ARGV.length < 1
    filename = ARGV[0]
    result = process(filename)
    puts "result = #{result}"
end

main if __FILE__ == $0

File: dart/2015/day03b/day03b.dart
#!/usr/bin/env dart

import 'dart:io';

import '../../utils/geometry/position2d.dart';

void usage(String progName) {
    print('usage: $progName <input file>');
    exit(1);
}

int process(String filename) {
    Position2D<int> santa = Position2D<int>(0, 0);
    Position2D<int> roboSanta = Position2D<int>(0, 0);
    var positions = <Position2D>{};
    bool santaMove = true;
    positions.add(santa);

    final lines = File(filename).readAsLinesSync();
    for ( final line in lines ) {
        for ( final ch in line.runes ) {
            if ( ch == '^'.codeUnitAt(0) ) {
                santaMove ? (santa = Position2D<int>(santa.x, santa.y + 1)) :
                    (roboSanta = Position2D<int>(roboSanta.x, roboSanta.y + 1));
            } else if ( ch == 'v'.codeUnitAt(0) ) {
                santaMove ? (santa = Position2D<int>(santa.x, santa.y - 1)) :
                    (roboSanta = Position2D<int>(roboSanta.x, roboSanta.y - 1));
            } else if ( ch == '<'.codeUnitAt(0) ) {
                santaMove ? (santa = Position2D<int>(santa.x - 1, santa.y)) :
                    (roboSanta = Position2D<int>(roboSanta.x - 1, roboSanta.y));
            } else if ( ch == '>'.codeUnitAt(0) ) {
                santaMove ? (santa = Position2D<int>(santa.x + 1, santa.y)) :
                    (roboSanta = Position2D<int>(roboSanta.x + 1, roboSanta.y));
            }
            santaMove ? positions.add(santa) : positions.add(roboSanta);
            santaMove = !santaMove;
        }
    }
    return positions.length;
}

void main(List<String> args) {
    if ( args.isEmpty ) {
        usage(Platform.script.pathSegments.last);
    }
    final filename = args[0];
    final result = process(filename);
    print('result = $result');
}
